x-pack/plugins/security_solution/public/detections/components/rules/rule_actions_field/index.tsx

```diff
@@ -98,20 +97,26 @@ export const RuleActionsField: React.FC<Props> = ({ field, messageVariables }) =
   const setActionParamsProperty = useCallback(
     // eslint-disable-next-line @typescript-eslint/no-explicit-any
     (key: string, value: any, index: number) => {
-      field.setValue((prevValue: RuleAction[]) => {
-        const updatedActions = [...prevValue];
-        updatedActions[index] = {
-          ...updatedActions[index],
-          params: {
-            ...updatedActions[index].params,
-            [key]: value,
-          },
-        };
-        return updatedActions;
-      });
+      // values in form controls and values used during form validation got out of sync leading to issue: https://github.com/elastic/kibana/issues/140593
```

Comment:
@vitaliidm Could you, please expand on why form and validation values get out of sync? If there's a race condition, using `setTimeout` to "win" it doesn't seem like a scalable solution.

Reply:
@xcrzx 

so, `setActionIdByIndex` within this file fires update of actions(but w/o params properties)
`setActionParamsProperty` updates params in field `actions` of form with values displayed if form controls through calling `field.setValue`. In our case, it is a message form field in slack action form.

These updates trigger form validation:

```typescript
  actions: {
    validations: [
      {
        validator: validateRuleActionsField(actionTypeRegistry),
      },
    ],
  },
```

`validateRuleActionsField`  receives in curried function argument field value:

```typescript
    const [{ value, path }] = data as [{ value: RuleAction[]; path: string }];
```

which doesn't have updates set in `setActionParamsProperty`. It happens consistently when you add some action, remove it, add again it. 

here is logged values in functions calls
```
setActionIdByIndex [
  {
    "id": "50a49d70-34d4-11ed-87dc-f3f665e2eedb",
    "actionTypeId": ".slack",
    "group": "default",
    "params": {}
  }
]
validateRuleActionsField [
  {
    "id": "50a49d70-34d4-11ed-87dc-f3f665e2eedb",
    "actionTypeId": ".slack",
    "group": "default",
    "params": {}
  }
]
validateRuleActionsField [
  {
    "id": "50a49d70-34d4-11ed-87dc-f3f665e2eedb",
    "actionTypeId": ".slack",
    "group": "default",
    "params": {}
  }
]
setActionParamsProperty [
  {
    "id": "50a49d70-34d4-11ed-87dc-f3f665e2eedb",
    "actionTypeId": ".slack",
    "group": "default",
    "params": {
      "message": "Rule {{context.rule.name}} generated {{state.signals_count}} alerts"
    }
  }
]

```

As you can see `validateRuleActionsField` is not getting called after `setActionParamsProperty` called, instead  validateRuleActionsField called twice on setActionIdByIndex. I would expect validation to be triggered once we set new params values.
But wrapping field.setValue in setTimeout helped to resolve the issue.
If you have any ideas how to resolve this, please share.

Reply:
> As you can see validateRuleActionsField is not getting called after setActionParamsProperty called, instead validateRuleActionsField called twice on setActionIdByIndex. I would expect validation to be triggered once we set new params values.

Thanks for the explanation! To me, that looks like a bug on the Actions Form side. Validation should definitely happen after params modification, but it's not the case. 

@elastic/response-ops Could you please confirm if the behavior described above is a bug? In which case, I propose fixing it in the `getActionForm` instead of working it around on the solution side.

Reply:
I believe the `validateRuleActionsField` validation occurs inside the security form, not inside the ActionForm. What triggers this validation?

Looking through the `triggers_actions_ui` code, when a connector type is specified, the `ActionForm` calls `setActionIdByIndex` with empty params and then it loads the Params component for that connector type. When the Params component is loaded, it sets a `defaultActionMessage` which calls `setActionParamsProperty`. Based on the logs above, the order that this is happening is correct. What is triggering the `validateRuleActionsField` call?

Reply:
> What is triggering the validateRuleActionsField call? 

I think it's triggered somewhere within [eui form hook ](https://github.com/elastic/kibana/blob/main/src/plugins/es_ui_shared/static/forms/hook_form_lib/hooks/use_form.ts)`useForm`, where `validateRuleActionsField` passed inside formSchema
```typescript
  actions: {
    validations: [
      {
        validator: validateRuleActionsField(actionTypeRegistry),
      },
    ],
  },
```

```typescript
  const { form } = useForm({
    schema: formSchema,
    defaultValue: defaultFormData,
  });

```

@xcrzx 
I think it would make sense to merge this fix and investigate further issue with not triggered validation. The root cause seems to be not in Security Solution, so we will depend on other teams in this, but current bug affects customer experience and we already on BC2 of 8.5 release.

cc: @banderror

Reply:
> I think it would make sense to merge this fix and investigate further issue with not triggered validation. The root cause seems to be not in Security Solution, so we will depend on other teams in this, but current bug affects customer experience and we already on BC2 of 8.5 release.

I'm fine with that. @vitaliidm, Could you please create a ticker for further investigation and mention it in the code comment for additional context?

Reply:
@xcrzx 
> I'm fine with that. @vitaliidm, Could you please create a ticker for further investigation and mention it in the code comment for additional context?

Here is created ticket https://github.com/elastic/kibana/issues/142217